src/main/kotlin/cacophony/utils/FileInput.kt

```diff
@@ -0,0 +1,91 @@
+package cacophony.utils
+
+import java.io.File
+
+class FileInput(filePath: String) : Input {
+    private val file = File(filePath)
+    private val lines = file.readLines().toMutableList()
+    private val lineBeginPos = mutableListOf<Int>()
+    private var curLineInd = 0
+    private var curPos = 0
+
+    init {
+        // Add new line character to every line (even if it might not have been present in the last one).
+        lines.forEachIndexed { ind, line ->
+            lines[ind] = line + "\n"
+        }
+
+        // Calculate position of each begin of line in the string formed by the concatenation of all lines.
+        // Used to encode Location.
+        var linesLengthSum = 0
+        lineBeginPos.add(0)
+        for (line in lines) {
+            linesLengthSum += line.length
+            lineBeginPos.add(linesLengthSum)
+        }
+    }
+
+    private fun setNextPosition() {
+        if (curLineInd >= lines.size) return
+
+        curPos++
+        if (curPos >= lines[curLineInd].length) {
+            curPos = 0
+            curLineInd++
+        }
+    }
+
+    private fun getCharAtPosition(
+        lineInd: Int,
+        pos: Int,
+    ): Char? {
+        if (lines.elementAtOrNull(lineInd) == null) return null
+        return lines[lineInd].elementAtOrNull(pos)
+    }
+
+    // Location is encoded as the position in the string formed by the concatenation of all lines.
+    private fun encodeLocation(
+        lineInd: Int,
+        pos: Int,
+    ): Location {
+        return Location(lineBeginPos[lineInd] + pos)
+    }
+
+    private fun decodeLocation(loc: Location): Pair<Int, Int> {
+        var lineInd = lineBeginPos.binarySearch(loc.value)
+        if (lineInd < 0) lineInd = -(lineInd + 2)
+        return Pair(lineInd, loc.value - lineBeginPos[lineInd])
+    }
+
+    override fun next(): Char? {
+        val nextChar = getCharAtPosition(curLineInd, curPos)
+        setNextPosition()
+        return nextChar
+    }
+
+    override fun peek(): Char? {
+        return getCharAtPosition(curLineInd, curPos)
+    }
+
+    override fun skip(c: Char) {
+        while ((peek() != null) and (peek() != c)) {
+            setNextPosition()
+        }
+    }
+
+    override fun getLocation(): Location {
+        return encodeLocation(curLineInd, curPos)
+    }
+
+    override fun setLocation(loc: Location) {
+        val (lineInd, pos) = decodeLocation(loc)
+        curLineInd = lineInd
+        curPos = pos
+    }
+
+    override fun locationToString(loc: Location): String {
```

Comment:
I would also add some function to print content between two locations or from one location to the end of the line? Would be helpful to show errors.